File: IOS_ApplePay/ApplePayObjc/ApplePayObjc/SQLModel/SqlManager.m
//
//  SqlManager.m
//  WildFireChat
//
//  Created by Hahn on 2020/6/9.
//  Copyright © 2020 WildFireChat. All rights reserved.
//

#import "SqlManager.h"
#import "ApplePayModel.h"
#import "FMDB.h"
#import "MJExtension.h"

static SqlManager *_instance = nil;

// Table SQL
/**
 *建表
 如果ApplePayModelTable这个表不存在就创建，表里面的字段有id、orderId、payData，并且orderId和payData不能为null，id为自增的主键
*/
static NSString *const kCreateStudentModelTableSQL = @"CREATE TABLE IF NOT EXISTS ApplePayModelTable(id integer primary key autoincrement,orderId text not null,payData text not null);";

// 表名
static NSString *const kApplePayModelTable = @"ApplePayModelTable";

// 字段名
static NSString *const kOrderIdColumn = @"orderId";
static NSString *const kPayDataColumn = @"payData";


@interface SqlManager()

@property (strong , nonatomic) FMDatabaseQueue * dbQueue;

@end

@implementation SqlManager

+ (instancetype)sharedManager {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _instance = [[SqlManager alloc] init];
        [_instance initial];
    });
    return _instance;
}

#pragma mark - ApplePayModel

- (void)saveApplePayModel:(ApplePayModel *)model{

    __block BOOL ret = NO;
    
    [_dbQueue inDatabase:^(FMDatabase * _Nonnull db) {
        NSError *error;
        ret = [db executeUpdate:[NSString stringWithFormat:@"INSERT INTO %@(%@,%@) VALUES (?,?);",kApplePayModelTable,kOrderIdColumn,kPayDataColumn] values:@[model.orderId, model.payData] error:&error];
        if (!ret) {
            NSLog(@"SQL保存失败 = %@",error.localizedDescription);
        } else {
            NSLog(@"SQL保存成功");
        }
    }];

}

- (NSArray<ApplePayModel *> *)getAllApplePayModel{
    __block NSMutableArray *modelArray = [[NSMutableArray alloc] init];
    
    [_dbQueue inDatabase:^(FMDatabase * _Nonnull db) {
        FMResultSet *set = [db executeQuery:[NSString stringWithFormat:@"SELECT * FROM %@",kApplePayModelTable]];
        while ([set next]) {
            
            int id = [set intForColumn:@"id"];
            NSString *orderId = [set stringForColumn:kOrderIdColumn];
            NSString *payData = [set stringForColumn:kPayDataColumn];
            NSDictionary *dict = [NSDictionary dictionaryWithObjectsAndKeys:[NSNumber numberWithInt:id], @"id", orderId, @"orderId", payData, @"payData", nil];
            ApplePayModel *model = [ApplePayModel mj_objectWithKeyValues:dict];
            if (model) {
                [modelArray insertObject:model atIndex:0];
            }
        }
    }];
    
    return modelArray.copy;
}


- (void)removeMotionModel:(ApplePayModel *)model {
    __block BOOL ret = NO;
    
    [_dbQueue inDatabase:^(FMDatabase * _Nonnull db) {
        NSError *error;
        ret = [db executeUpdate:[NSString stringWithFormat:@"DELETE FROM %@ WHERE %@ = ?",kApplePayModelTable,kOrderIdColumn] values:@[model.orderId] error:&error];
        if (!ret) {
            NSLog(@"删除SQL失败 = %@",error);
        }
    }];

}

#pragma mark - Primary

- (void)initial {
    __unused BOOL result = [self initDatabase];
    NSAssert(result, @"Init Database fail");
}

/**
 * 初始化SQL 保存地址的操作
 */
- (BOOL)initDatabase {
    NSString *doc = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) lastObject];
    doc = [doc stringByAppendingPathComponent:@"LDebugTool"];
    if (![[NSFileManager defaultManager] fileExistsAtPath:doc]) {
        NSError *error;
        [[NSFileManager  defaultManager] createDirectoryAtPath:doc withIntermediateDirectories:YES attributes:nil error:&error];
        if (error) {
            NSLog(@"CSStorageManager create folder fail, error = %@",error.description);
        }
        NSAssert(!error, error.description);
    }
    NSString *filePath = [doc stringByAppendingPathComponent:@"LDebugTool.db"];
    
    _dbQueue = [FMDatabaseQueue databaseQueueWithPath:filePath];
    
    __block BOOL ret1 = NO;
    [_dbQueue inDatabase:^(FMDatabase * _Nonnull db) {
        // ObjectData use to convert to Student, launchDate use as identity
        ret1 = [db executeUpdate:kCreateStudentModelTableSQL];
        if (!ret1) {
            NSLog(@"LLStorageManager create StudentModelTable fail");
        }
    }];
    return ret1;
}

@end

File: IOS_ApplePay/ApplePaySwift/ApplePaySwift/SQLModel/SqlManager.swift
//
//  SqlManager.swift
//  ApplePaySwift
//
//  Created by Hahn on 2020/6/10.
//  Copyright © 2020 Hahn. All rights reserved.
//

import UIKit

class SqlManager: NSObject {
    
    // MARK: -单例
    private static var sharedSqlManager: SqlManager = {
        let sqlManager = SqlManager()
        return sqlManager
    }()

    override init() {
        var doc:NSString! = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).last! as NSString
        doc = doc.appendingPathComponent("LDebugTool") as NSString
        
        if !FileManager.default.fileExists(atPath: doc! as String) {
            do {
                try FileManager.default.createDirectory(atPath: doc! as String, withIntermediateDirectories: true, attributes: nil)
            } catch  {
                print(error)
            }
        }
        
        let filePath:NSString! = doc.appendingPathComponent("LDebugTool.db") as NSString
        self.dbQueue = FMDatabaseQueue(path: filePath as String)
        

        let sql:NSString! = self.kCreateStudentModelTableSQL as NSString
        self.dbQueue?.inDatabase { (db) in
            
            let result = db.executeUpdate(sql as String, withArgumentsIn: [])
            if result {
                print("创表成功")
            }else{
                print("创表失败")
            }
        }
        
    }

    class func shared() -> SqlManager {
        return sharedSqlManager
    }
    
    private var dbQueue:FMDatabaseQueue?
    
    /**
     *建表
     如果ApplePayModelTable这个表不存在就创建，表里面的字段有id、orderId、payData，并且orderId和payData不能为null，id为自增的主键
    */
    private let kCreateStudentModelTableSQL:String! = "CREATE TABLE IF NOT EXISTS ApplePayModelTable(id integer primary key autoincrement,orderId text not null,payData text not null);"
    
    // 表名
    private let kApplePayModelTable:String! = "ApplePayModelTable"

    // 字段名
    private let kOrderIdColumn:String! = "orderId"
    private let kPayDataColumn:String! = "payData"
    
    // MARK:- ApplePayModel
    /**
     * 保存数据到SQL
     */
    public func saveApplePayModel(model:ApplePayModel){
        let table:String! = self.kApplePayModelTable
        let order:String! = self.kOrderIdColumn
        let data:String! = self.kPayDataColumn
        
        self.dbQueue?.inDatabase({ (db) in
            do {
                try db.executeUpdate("INSERT INTO \(table!)(\(order!),\(data!)) VALUES (?,?);", values: [model.orderId as Any,model.payData as Any])
            } catch  {
                print(error)
            }
            
        })
    }

    /**
     * 读取全部数据
     */
    public func getAllApplePayModel() -> NSMutableArray{
        let table:String! = self.kApplePayModelTable
        let array:NSMutableArray! = NSMutableArray()
        self.dbQueue?.inDatabase({ (db) in
            do {
                let set:FMResultSet = try db.executeQuery("SELECT * FROM \(table!)", values: nil)
                
                while set.next() {
                    _ = set.int(forColumn: "id")
                    let orderId = set.string(forColumn: "orderId")
                    let payData = set.string(forColumn: "payData")

                    let model = ApplePayModel()
                    model.orderId = orderId
                    model.payData = payData
                    array.add(model)
                    
                }
            } catch  {
                print(error)
            }
            
        })
        return array
    }

    /**
     * 删除指定的数据
     */
    public func removeMotionModel(model:ApplePayModel){
        let table:String! = self.kApplePayModelTable
        let order:String! = self.kOrderIdColumn
        
        self.dbQueue?.inDatabase({ (db) in
            do {
                try db.executeUpdate("DELETE FROM \(table!) WHERE \(order!) = ?", values: [model.orderId as Any])
            } catch  {
                print(error)
            }
            
        })
    }
}
